### DISCUSSION TITLE
Automatically scale resolution of image overlay

### DISCUSSION THREAD

**Discussion by ero-qt** [Leaflet/Leaflet#8792] [Q&A] [answered]
Hi. I'm using an `imageOverlay` for a map of a video game. This map is currently very large; 8192 x 8192. This is obviously very bad for performance on the site.

I was wondering, is it possible to scale this image automatically in the map container, or perhaps provide different resolutions? I've tried a `tileLayer` approach, but I find it unreasonable to cut my map into 256 x 256 pixel large chunks at 10 different resolutions. That's multiple thousand tiles. I'm not doing that manually. And even if I did do that, the `tileLayer` version has that wrapping mechanism, which isn't applicable to game maps.

What can I do here?

**Comment by IvanSanchez** [comment, marked as answer]
You'll want tiles anyway. Tiles cover the "different resolutions" problem, and the "not loading data that is not on the screen" problem. Besides, a 8192x8192px raster is just 1024 tiles - not too many.

> I'm not doing that manually.

That's fine! Use https://github.com/jahed/maptiles , or https://gdal.org/programs/gdal2tiles.html , or imagemagick, or zoomify, or whatever tool you want!

> And even if I did do that, the tileLayer version has that wrapping mechanism, which isn't applicable to game maps.

Read the `L.CRS.Simple` tutorial, and use the `maxBounds` option in your `TileLayer`.

**Reply by ero-qt** [reply]
Getting somewhere. Had to install WSL2 on my windows machine to get those tools to work, but eventually I managed to generate my tiles. Thanks for that.

Using `L.CRS.Simple` did remove the wrapping, but brought different issues.  
Now, according to the documentation, there is no `maxBounds` option on `TileLayer`. There is `bounds` on `GridLayer` (which `TileLayer` inherits from) and there's `maxBounds` on `Map`. Which one did you mean? Neither of them appears to do anything for me. If you'd like, you can check my work:

```js
// Using the original image's size as the bounds.
const bounds = [[0, 0], [8192, 8192]];
const map = L.map("map", {
  crs: L.CRS.Simple,
  // maxBounds: bounds // This was the first attempt. Didn't do anything.
}).setView([0, 0], 2); // Unrelated, but is there a way to automatically center the map?

L.tileLayer("tiles/{z}/tile_{x}_{y}.png", {
  minZoom: 1,
  maxZoom: 5,
  // bounds: bounds // This was the second attempt. Didn't do anything either.
}).addTo(map);

// I also tried `map.fitBounds(bounds);` here, but that didn't change anything.
```

**Reply by ero-qt** [reply]
Looks like I got it to work the way I wanted it to.

```js
const TILE_SIZE = 256;

const bounds = [[0, 0], [-TILE_SIZE, TILE_SIZE]];

const map = L.map("map", {
  crs: L.CRS.Simple,
  maxBounds: bounds,
  tileSize: TILE_SIZE
}).setView([0, 0], 1);

L.tileLayer("assets/maps/tiles/{z}/tile_{x}_{y}.png", {
  minZoom: 1,
  maxZoom: 5,
  updateInterval: 10
}).addTo(map);
```

There are a couple other questions I have. There is a secondary map I'm using which basically removes all foliage from the game. I would like to switch these maps on the click of a button. Does LeafletJS have something like this built in?  
I would also like to use a slider to do the zooming. Can the map's zoom level be adjusted between the zoom levels? Can the `+` and `-` buttons in the top-left be disabled?

**Reply by IvanSanchez** [reply]
> There is `bounds` on `GridLayer` (which `TileLayer` inherits from)

...which means that there's `bounds` on `TileLayer`. That's the one I meant.

> Neither of them appears to do anything for me.

`maxBounds` on `L.Map` will prevent the user from going outside that area; `bounds` on `GridLayer` will prevent tiles outside that area from being loaded. For the latter case, do check the network tab of your browser's developer tools: you should see `404` errors when there's no `bounds` of your `TileLayer` and you're viewing an area outside of those bounds.

> I would like to switch these maps on the click of a button. 

Read the tutorial on `L.Control.Layers`.

> I would also like to use a slider to do the zooming.

Read https://leafletjs.com/plugins#interactive-panzoom

> Can the + and - buttons in the top-left be disabled?

See https://leafletjs.com/reference.html#map-zoomcontrol

**Reply by ero-qt** [reply]
Neither `bounds: [[0, 0], [8192, 8192]]` (the size of the image), nor `bounds: [[0, 0], [31, 31]]` (the maximum x and y tiles) appear to work. The map either doesn't render, or renders off-screen. I'm not sure. The 404 errors still appear in both cases. I did also try with `Map.fitBounds` with both versions.

Additionally to this, there appears to be some... "weirdness" when dragging beyond the *map*'s bounds, when it attempts to snap back into place. Dragging becomes very strange and inconsistent, sometimes not responding at all, other times dragging into the complete opposite direction of what was input.
